README
NAME
    Mojo::IRC - IRC Client for the Mojo IOLoop

VERSION
    0.43

SYNOPSIS
      my $irc = Mojo::IRC->new(
                  nick => 'test123',
                  user => 'my name',
                  server => 'irc.perl.org:6667',
                );

      $irc->on(irc_join => sub {
        my($self, $message) = @_;
        warn "yay! i joined $message->{params}[0]";
      });

      $irc->on(irc_privmsg => sub {
        my($self, $message) = @_;
        say $message->{prefix}, " said: ", $message->{params}[1];
      });

      $irc->connect(sub {
        my($irc, $err) = @_;
        return warn $err if $err;
        $irc->write(join => '#mojo');
      });

      Mojo::IOLoop->start;

DESCRIPTION
    Mojo::IRC is a non-blocking IRC client using Mojo::IOLoop from the
    wonderful Mojolicious framework.

    It features IPv6 and TLS, with additional optional modules:
    IO::Socket::IP and IO::Socket::SSL.

    By default this module will only emit standard IRC events, but by
    settings "parser" to a custom object it will also emit CTCP events.
    Example:

      my $irc = Mojo::IRC->new;
      $irc->parser(Parse::IRC->new(ctcp => 1);
      $irc->on(ctcp_action => sub {
        # ...
      });

    It will also set up some default events: "ctcp_ping", "ctcp_time", and
    "ctcp_version".

    This class inherits from Mojo::EventEmitter.

TESTING
    The module Test::Mojo::IRC is useful if you want to write tests without
    having a running IRC server.

    MOJO_IRC_OFFLINE (from v0.20) is now DEPRECATED in favor of
    Test::Mojo::IRC.

EVENTS
  close
      $self->on(close => sub { my ($self) = @_; });

    Emitted once the connection to the server closes.

  error
      $self->on(error => sub { my ($self, $err) = @_; });

    Emitted once the stream emits an error.

  message
      $self->on(message => sub { my ($self, $msg) = @_; });

    Emitted when a new IRC message arrives. Will dispatch to a default
    handler, which will again emit "err_event_name" "ctcp_event_name" and
    "irc_event_name" below.

    Here is an example $msg:

      {
        command  => "PRIVMSG",
        event    => "privmsg",
        params   => ["#convos", "hey!"],
        prefix   => "jan_henning",
        raw_line => ":jan_henning PRIVMSG #convos :hey",
      }

  err_event_name
    Events that start with "err_" are emitted when there is an IRC response
    that indicates an error. See Mojo::IRC::Events for sample events.

  ctcp_event_name
    Events that start with "ctcp_" are emitted if the "parser" can
    understand CTCP messages, and there is a CTCP response.

      $self->parser(Parse::IRC->new(ctcp => 1);

    See Mojo::IRC::Events for sample events.

  irc_event_name
    Events that start with "irc_" are emitted when there is a normal IRC
    response. See Mojo::IRC::Events for sample events.

ATTRIBUTES
  connect_timeout
      $int = $self->connect_timeout;
      $self = $self->connect_timeout(60);

    Maximum amount of time in seconds establishing a connection may take
    before getting canceled, defaults to the value of the
    "MOJO_IRC_CONNECT_TIMEOUT" environment variable or 30.

  ioloop
    Holds an instance of Mojo::IOLoop.

  local_address
      $str = $self->local_address;
      $self = $self->local_address("10.20.30.40");

    Local address to bind to. See "local_address" in Mojo::IOLoop::Client.

  name
    The name of this IRC client. Defaults to "Mojo IRC".

  nick
    IRC nick name accessor. Default to "user".

  parser
      $self = $self->parser($obj);
      $self = $self->parser(Parse::IRC->new(ctcp => 1));
      $obj = $self->parser;

    Holds a Parse::IRC object by default.

  pass
    Password for authentication

  real_host
    Will be set by "irc_rpl_welcome". Holds the actual hostname of the IRC
    server that we are connected to.

  server
    Server name and, optionally, a port to connect to. Changing this while
    connected to the IRC server will issue a reconnect.

  server_settings
      $hash = $self->server_settings;

    Holds information about the server. See
    <https://github.com/jhthorsen/mojo-irc/blob/master/t/ua-channel-users.t>
    for example data structure.

    Note that this attribute is EXPERIMENTAL and the structure of the values
    it holds.

  user
    IRC username. Defaults to current logged in user or falls back to
    "anonymous".

  tls
      $self->tls(undef) # disable (default)
      $self->tls({}) # enable

    Default is "undef" which disables TLS. Setting this to an empty hash
    will enable TLS and this module will load in default certs. It is also
    possible to set custom cert/key:

      $self->tls({ cert => "/path/to/client.crt", key => ... })

    This can be generated using

      # certtool --generate-privkey --outfile client.key
      # certtool --generate-self-signed --load-privkey client.key --outfile client.crt

METHODS
  connect
      $self = $self->connect(\&callback);

    Will log in to the IRC "server" and call &callback. The &callback will
    be called once connected or if connect fails. The second argument will
    be an error message or empty string on success.

  ctcp
      $str = $self->ctcp(@str);

    This message will quote CTCP messages. Example:

      $self->write(PRIVMSG => nickname => $self->ctcp(TIME => time));

    The code above will write this message to IRC server:

      PRIVMSG nickname :\001TIME 1393006707\001

  disconnect
      $self->disconnect(\&callback);

    Will disconnect form the server and run the callback once it is done.

  new
      $self = Mojo::IRC->new(%attrs);

    Object constructor.

  register_default_event_handlers
      $self->register_default_event_handlers;

    This method sets up the default "DEFAULT EVENT HANDLERS" unless someone
    has already subscribed to the event.

  write
      $self->write(@str, \&callback);

    This method writes a message to the IRC server. @str will be
    concatenated with " " and "\r\n" will be appended. &callback is called
    once the message is delivered over the stream. The second argument to
    the callback will be an error message: Empty string on success and a
    description on error.

DEFAULT EVENT HANDLERS
  ctcp_ping
    Will respond to the sender with the difference in time.

      Ping reply from $sender: 0.53 second(s)

  ctcp_time
    Will respond to the sender with the current localtime. Example:

      TIME Fri Feb 21 18:56:50 2014

    NOTE! The localtime format may change.

  ctcp_version
    Will respond to the sender with:

      VERSION Mojo-IRC $VERSION

    NOTE! Additional information may be added later on.

  irc_nick
    Used to update the "nick" attribute when the nick has changed.

  irc_notice
    Responds to the server with "QUOTE PASS ..." if the notice contains
    "Ident broken...QUOTE PASS...".

  irc_ping
    Responds to the server with "PONG ...".

  irc_rpl_isupport
    Used to populate "server_settings" with information about the server.

  irc_rpl_welcome
    Used to get the hostname of the server. Will also set up automatic PING
    requests to prevent timeout and update the "nick" attribute.

  err_nicknameinuse
    This handler will add "_" to the failed nick before trying to register
    again.

COPYRIGHT
    This program is free software, you can redistribute it and/or modify it
    under the terms of the Artistic License version 2.0.

AUTHOR
    Marcus Ramberg - "mramberg@cpan.org"

    Jan Henning Thorsen - "jhthorsen@cpan.org"



Changes
Revision history for perl distribution Mojo-IRC

0.47 Not Released
 - Removed deprecated method track_any()
 - Passthrough ctcpreply events from Parse::IRC #32
   Contributor: Dominik Hassler
 - Text for NOTICE is in the second param #33
   Contributor: Dan Book

0.46 2020-02-16
 - Fix insecure option for Mojolicious 9.0
 - Add write_p method returning a Mojo::Promise
 - Use promises instead of deprecated and decored Mojo::IOLoop::Delay
 - Deprecated Mojo::IRC::UA

0.45 2018-06-07T14:41:06+0800
 - Add insecure option for invalid certs with Mojolicious 7.80
   Contributor: Dan Book

0.44 2017-02-18T02:50:18+0100
 - Fix Mojo::IRC::UA did not clean up events

0.43 2017-01-14T11:10:32+0100
 - Add "message" event #29
 - Replaced EXPERIMENTAL "irc_any" event with "message" event #29
 - Compatible with Mojolicious 7.18 #29

0.42 2016-11-16T20:04:49+0100
 - Can detect server settings using rpl_isupport()
 - Improved parsing nicks with modes Nordaaker/Convos#287

0.41 2016-11-13T14:01:01+0100
 - Fix parsing nicks with & and % modes Nordaaker/Convos#287

0.40 2016-10-13T09:40:46+0200
 - Add support for JOIN redirects Nordaaker/Convos#284

0.39 2016-09-16T09:34:40+0200
 - Add "host" and "server_info" to Mojo::IRC::UA->whois() output

0.38 2016-09-14T23:10:59+0200
 - Fix matching channel names case insensitive in Mojo::IRC::UA

0.37 2016-09-07T09:17:08+0200
 - Skipping t/ua-mode.t for now. https://github.com/jhthorsen/mojo-irc/issues/28

0.36 2016-09-06T21:31:13+0200
 - Fix t/ua-mode.t #28

0.35 2016-08-29T09:14:48+0200
 - Add Mojo::IRC::UA->kick() method
 - Add Mojo::IRC::UA->mode() method

0.34 2016-08-25T22:44:57+0200
 - Add connect_timeout() attribute

0.33 2016-08-24T22:40:14+0200
 - Fix Mojo::IRC::UA triggered callback on wrong IRC event

0.32 2016-08-24T17:15:41+0200
 - Fix joining channel with key

0.31 2016-04-09T23:49:41+0200
 - Add Test::Mojo::IRC->server attribute
 - Add Test::Mojo::IRC->start_server can be called as a class method
 - Change channel_topic() to return {topic => "..."}
 - Fix channel_topic() return value on error

0.30 2016-03-22T12:00:55+0100
 - Fix channels() on error to return hash-ref
 - Breaking change: "irc_error" is not always emitted
 - Add EXPERIMENTAL "irc_any" event

0.29 2016-03-10T22:22:24+0100
 - Prevent t/parse-irc-message.t from getting run

0.28 2016-02-28T13:48:40+0100
 - Add Mojo::IRC::UA->part_channel() method
 - Add "local_address" attribute

0.27 2015-09-21T11:22:24+0200
 - Fix calling callbacks async
 - Remove support for MOJO_IRC_OFFLINE=1

0.26 2015-07-11T13:26:26+0200
 - Bumping Mojolicious requirement to 6.00

0.25 2015-06-15T15:31:27+0200
 - Remove [+nt] and friends from topic returned from channels()
 - Fix parsing to allow CR or LF at buffer end. #24
   Contributor: Ben Tyler

0.24 2015-06-02T22:18:12+0200
 - Fix default nick only contains [a-z] and "_".

0.23 2015-06-02T00:46:24+0200
 - Fix requirement to Mojo::Util (Need version 1.33)

0.22 2015-05-31T09:21:58+0200
 - Add Mojo::IRC::UA
 - Add Mojo::IRC::UA->channels() method
 - Add Mojo::IRC::UA->channel_topic() method
 - Add Mojo::IRC::UA->channel_users() method
 - Add Mojo::IRC::UA->join_channel() method
 - Add Mojo::IRC::UA->nick() method
 - Add Mojo::IRC::UA->whois() method

0.21 2015-05-31T07:54:33+0200
 - Deprecated MOJO_IRC_OFFLINE in favor of Test::Mojo::IRC

0.20 2015-05-30T19:29:25+0200
 - Test::Mojo::IRC can fetch server output from __DATA__ section
 - Fix "choppy" buffer from Test::Mojo::IRC test server

0.19 2015-05-22T13:44:50+0200
 - Add Test::Mojo::IRC
 - Changed default value for nick() to mirror user()
 - Changed default value for user() to reflect current logged in user

0.18 2015-03-01T18:05:08Z
 - More tests are skipped on MSWin32

0.17 2015-02-15T21:44:25Z
 - Fix parsing bug in _read() on unknown numeric IRC event, #21
 - Fix setting nick in irc_rpl_welcome default handler, #20

0.16 2015-02-08T11:41:48Z
 - Add disconnect() can be called without a callback Contributor: Dan Book
 - Fix "close" event can reconnect Contributor: Dan Book
 - Fix responding to CTCP messages
 - Fix responding to CTCP PING message

0.15 2015-01-09T16:19:34Z
 - More robust unit tests

0.14 2015-01-02T13:31:11Z
 - Fix failing test on BSD

0.13 2014-09-14T20:16:25Z
 - emit_safe() is deprecated in Mojolicious

0.12 2014-09-06T19:25:13Z
 - Skip tests on MSWin32 #9

0.11 2014-08-31T14:32:11Z
 - Fix update of current nick() on err_nicknameinuse()

0.10 2014-08-16T20:24:25Z
 - Fix update of current nick()
 - Deprecated change_nick()

0.09 2014-07-29T08:43:34Z
 - Fix emitting err_xxx events before irc_error

0.08 2014-07-27T19:06:53Z
 - Will also emit numeric events
 - Move events out of main module documentation
 - Deprecated irc_err_xxx events

0.07 2014-05-30T07:46:58Z
 - Fix Mojo::IRC for Mojolicious 5.0

0.0601 2014-02-22T17:43:57Z
 - Fix missing ERR_ events.
 - Started DEPRECATION process for irc_err_xxx events.

0.06 2014-02-22T11:54:38Z
 - Allow offline mode with MOJO_IRC_OFFLINE=1

0.0501 2014-02-22T09:57:25Z
 - Add debug info to figure out failing error.t
   http://www.cpantesters.org/cpan/report/3758769c-6bf4-1014-9dd5-8faa6c3f6
   a79

0.05 2014-02-21T20:52:39Z
 - Add parser() attribute
 - Add ctcp() method to quote CTCP messages
 - Add support for CTCP events

0.04 2014-02-21T18:23:16Z
 - Fix PASS need to be sent before NICK and USER
   http://www.irchelp.org/irchelp/rfc/chapter4.html#c4_1_1 Contributor:
   Stephan Jauernick

0.0303 2013-10-01T12:23:30Z
 - Fix irc_err_nicknameinuse: Need to try with same nick, not modify the
   nick attribute.
 - Marked change_nick() as deprecated.

0.0302 2013-09-06T23:40:46Z
 - Fix t/error.t on windows
 - Fix Makefile.PL

0.0301 2013-09-04T23:46:41Z
 - Need to remove stream_id on error as well

0.03 2013-09-04T13:48:26Z
 - Add support for TLS
 - Change default ping interval to every 60 seconds

0.0201 2013-08-16T23:31:03Z
 - Fix NAME in Makefile.PL

0.02 2013-08-05T12:40:20Z
 - Remove File::Slurp as a dependency
 - Remove Carp as a dependency

t/Helper.pm
package t::Helper;
use Mojo::Util 'monkey_patch';
use Test::More ();
use Mojo::IRC;

sub import {
  my $class  = shift;
  my $caller = caller;

  eval <<"HERE" or die $@;
  package $caller;
  use Test::More;
  use strict;
  use warnings;
  1;
HERE

  eval "use Mojo::IOLoop; use Mojo::IOLoop::Server";

  monkey_patch $caller => generate_port => sub {
    Mojo::IOLoop->can('generate_port') ? Mojo::IOLoop->generate_port : Mojo::IOLoop::Server->generate_port;
  };
}

1;


Makefile.PL
# Generated by git-ship. See 'git-ship --man' for help or https://github.com/jhthorsen/app-git-ship
use ExtUtils::MakeMaker;
my %WriteMakefileArgs = (
  NAME           => 'Mojo::IRC',
  AUTHOR         => 'Jan Henning Thorsen <jhthorsen@cpan.org>',
  LICENSE        => 'artistic_2',
  ABSTRACT_FROM  => 'lib/Mojo/IRC.pm',
  VERSION_FROM   => 'lib/Mojo/IRC.pm',
  EXE_FILES      => [qw()],
  BUILD_REQUIRES => {},
  TEST_REQUIRES  => {'Test::More' => '0.88'},
  PREREQ_PM      => {
    'IRC::Utils'    => '0.12',
    'List::Util'    => '1.33',
    'Mojolicious'   => '7.18',
    'Parse::IRC'    => '1.20',
    'Unicode::UTF8' => '0.59'
  },
  META_MERGE => {
    'dynamic_config' => 0,
    'meta-spec'      => {version => 2},
    'resources'      => {
      bugtracker => {web => 'https://github.com/jhthorsen/mojo-irc/issues'},
      homepage   => 'https://github.com/jhthorsen/mojo-irc',
      repository => {
        type => 'git',
        url  => 'https://github.com/jhthorsen/mojo-irc.git',
        web  => 'https://github.com/jhthorsen/mojo-irc',
      },
    },
  },
  test => {S => (-e 'META.yml' ? 't/*.t' : 't/*.t xt/*.t')},
);

unless (eval { ExtUtils::MakeMaker->VERSION('6.63_03') }) {
  my $test_requires = delete $WriteMakefileArgs{TEST_REQUIRES};
  @{$WriteMakefileArgs{PREREQ_PM}}{keys %$test_requires} = values %$test_requires;
}

WriteMakefile(%WriteMakefileArgs);


lib/Mojo/IRC.pm
package Mojo::IRC;
use Mojo::Base 'Mojo::EventEmitter';
use Mojo::IOLoop;
use Mojo::Promise;
use File::Basename 'dirname';
use File::Spec::Functions 'catfile';
use IRC::Utils   ();
use Parse::IRC   ();
use Scalar::Util ();
use Unicode::UTF8;
use constant DEBUG        => $ENV{MOJO_IRC_DEBUG}     || 0;
use constant DEFAULT_CERT => $ENV{MOJO_IRC_CERT_FILE} || catfile dirname(__FILE__), 'mojo-irc-client.crt';
use constant DEFAULT_KEY  => $ENV{MOJO_IRC_KEY_FILE}  || catfile dirname(__FILE__), 'mojo-irc-client.key';

our $VERSION = '0.47';

our %NUMERIC2NAME = (470 => 'ERR_LINKCHANNEL');

my %CTCP_QUOTE = ("\012" => 'n', "\015" => 'r', "\0" => '0', "\cP" => "\cP");

my @DEFAULT_EVENTS = qw(
  irc_ping irc_nick irc_notice irc_rpl_welcome err_nicknameinuse
  irc_rpl_isupport ctcp_ping ctcp_time ctcp_version
);

has connect_timeout => sub { $ENV{MOJO_IRC_CONNECT_TIMEOUT} || 30 };
has ioloop          => sub { Mojo::IOLoop->singleton };
has local_address   => '';
has name            => 'Mojo IRC';
has nick            => sub { shift->_build_nick };
has parser          => sub { Parse::IRC->new; };
has pass            => '';
has real_host       => '';

has server_settings => sub {
  return {chantypes => '#', prefix => '(ov)@+'};
};

has tls  => undef;
has user => sub { $ENV{USER} || getlogin || getpwuid($<) || 'anonymous' };

sub new {
  my $self = shift->SUPER::new(@_);
  $self->on(message => \&_legacy_dispatch_message);
  return $self;
}

sub server {
  my ($self, $server) = @_;
  my $old = $self->{server} || '';

  Scalar::Util::weaken($self);
  return $old unless defined $server;
  return $self if $old and $old eq $server;
  $self->{server} = $server;
  return $self unless $self->{stream_id};
  $self->disconnect(sub {
    $self->connect(sub { });
  });
  $self;
}

sub connect {
  my ($self, $cb) = @_;
  my ($host, $port) = split /:/, $self->server;
  my @extra;

  if (!$host) {
    $self->ioloop->next_tick(sub { $self->$cb('server() is not set.') });
    return $self;
  }
  if ($self->{stream_id}) {
    $self->ioloop->next_tick(sub { $self->$cb('') });
    return $self;
  }

  if ($self->local_address) {
    push @extra, local_address => $self->local_address;
  }
  if (my $tls = $self->tls) {
    push @extra, tls         => 1;
    push @extra, tls_ca      => $tls->{ca} if $tls->{ca};  # not sure why this should be supported, but adding it anyway
    push @extra, tls_cert    => $tls->{cert} || DEFAULT_CERT;
    push @extra, tls_key     => $tls->{key}  || DEFAULT_KEY;
    push @extra, tls_verify  => 0x00                      if $tls->{insecure};    # Mojolicious < 9.0
    push @extra, tls_options => {SSL_verify_mode => 0x00} if $tls->{insecure};    # Mojolicious >= 9.0
  }

  $port ||= 6667;
  $self->{buffer} = '';
  $self->{debug_key} ||= "$host:$port";
  $self->register_default_event_handlers;

  Scalar::Util::weaken($self);
  $self->{stream_id} = $self->ioloop->client(
    address => $host,
    port    => $port,
    timeout => $self->connect_timeout,
    @extra,
    sub {
      my ($loop, $err, $stream) = @_;

      if ($err) {
        delete $self->{stream_id};
        return $self->$cb($err);
      }

      $stream->timeout(0);
      $stream->on(
        close => sub {
          $self or return;
          warn "[$self->{debug_key}] : close\n" if DEBUG;
          delete $self->{stream};
          delete $self->{stream_id};
          $self->emit('close');
        }
      );
      $stream->on(
        error => sub {
          $self         or return;
          $self->ioloop or return;
          $self->ioloop->remove(delete $self->{stream_id});
          $self->emit(error => $_[1]);
        }
      );
      $stream->on(read => sub { $self->_read($_[1]) });

      $self->{stream} = $stream;
      $self->ioloop->next_tick(sub {
        my @promises;
        push @promises, $self->write_p(PASS => $self->pass) if length $self->pass;
        push @promises, $self->write_p(NICK => $self->nick);
        push @promises, $self->write_p(USER => $self->user, 8, '*', ':' . $self->name);
        Mojo::Promise->all(@promises)->finally(sub { $self->$cb('') });
      });
    }
  );

  return $self;
}

sub ctcp {
  my $self = shift;
  local $_ = join ' ', @_;
  s/([\012\015\0\cP])/\cP$CTCP_QUOTE{$1}/g;
  s/\001/\\a/g;
  ":\001${_}\001";
}

sub disconnect {
  my ($self, $cb) = @_;

  if (my $tid = delete $self->{ping_tid}) {
    $self->ioloop->remove($tid);
  }

  if ($self->{stream}) {
    Scalar::Util::weaken($self);
    $self->{stream}->write(
      "QUIT\r\n",
      sub {
        $self->{stream}->close;
        $self->$cb if $cb;
      }
    );
  }
  elsif ($cb) {
    $self->ioloop->next_tick(sub { $self->$cb });
  }

  $self;
}

sub register_default_event_handlers {
  my $self = shift;

  for my $event (@DEFAULT_EVENTS) {
    $self->on($event => $self->can($event)) unless $self->has_subscribers($event);
  }

  return $self;
}

sub write {
  no warnings 'utf8';
  my $cb   = ref $_[-1] eq 'CODE' ? pop : sub { };
  my $self = shift;
  my $buf  = Unicode::UTF8::encode_utf8(join(' ', @_), sub { $_[0] });

  Scalar::Util::weaken($self);
  if (ref $self->{stream}) {
    warn "[$self->{debug_key}] <<< $buf\n" if DEBUG;
    $self->{stream}->write("$buf\r\n", sub { $self->$cb(''); });
  }
  else {
    $self->ioloop->next_tick(sub { $self->$cb('Not connected.') });
  }

  $self;
}

sub write_p {
  my ($self, @args) = @_;
  my $p = Mojo::Promise->new->ioloop($self->ioloop);
  $self->write(@args, sub { length $_[1] ? $p->reject($_[1]) : $p->resolve(1) });
  return $p;
}

sub ctcp_ping {
  my ($self, $message) = @_;
  my $ts   = $message->{params}[1];
  my $nick = IRC::Utils::parse_user($message->{prefix});

  return $self unless $ts;
  return $self->write('NOTICE', $nick, $self->ctcp(PING => $ts));
}

sub ctcp_time {
  my ($self, $message) = @_;
  my $nick = IRC::Utils::parse_user($message->{prefix});

  $self->write(NOTICE => $nick, $self->ctcp(TIME => scalar localtime));
}

sub ctcp_version {
  my ($self, $message) = @_;
  my $nick = IRC::Utils::parse_user($message->{prefix});

  $self->write(NOTICE => $nick, $self->ctcp(VERSION => 'Mojo-IRC', $VERSION));
}

sub irc_nick {
  my ($self, $message) = @_;
  my $old_nick = ($message->{prefix} =~ /^[~&@%+]?(.*?)!/)[0] || '';

  if (lc $old_nick eq lc $self->nick) {
    $self->nick($message->{params}[0]);
  }
}

sub irc_notice {
  my ($self, $message) = @_;

  # NOTICE AUTH :*** Ident broken or disabled, to continue to connect you must type /QUOTE PASS 21105
  if ($message->{params}[1] =~ m!Ident broken.*QUOTE PASS (\S+)!) {
    $self->write(QUOTE => PASS => $1);
  }
}

sub irc_ping {
  my ($self, $message) = @_;
  $self->write(PONG => $message->{params}[0]);
}

sub irc_rpl_isupport {
  my ($self, $message) = @_;
  my $params          = $message->{params};
  my $server_settings = $self->server_settings;
  my %got;

  for my $i (1 .. @$params - 1) {
    next unless $params->[$i] =~ /([A-Z]+)=?(\S*)/;
    my ($k, $v) = (lc $1, $2);
    $got{$k} = 1;
    $server_settings->{$k} = $v || 1;
  }
}

sub irc_rpl_welcome {
  my ($self, $message) = @_;
  $self->nick($message->{params}[0]);

  Scalar::Util::weaken($self);
  $self->real_host($message->{prefix});
  $self->{ping_tid} ||= $self->ioloop->recurring(
    $self->{ping_pong_interval} || 60,    # $self->{ping_pong_interval} is EXPERIMENTAL
    sub {
      $self->write(PING => $self->real_host);
    }
  );
}

sub err_nicknameinuse {
  my ($self, $message) = @_;
  my $nick = $message->{params}[1];

  $self->nick($nick . '_');
  $self->write(NICK => $self->nick, sub { });
}

sub DESTROY {
  my $self   = shift;
  my $ioloop = $self->ioloop or return;
  my $tid    = $self->{ping_tid};
  my $sid    = $self->{stream_id};

  $ioloop->remove($sid) if $sid;
  $ioloop->remove($tid) if $tid;
}

sub _build_nick {
  my $nick = shift->user;
  $nick =~ s![^a-z_]!_!g;
  $nick;
}

sub _dispatch_message {
  my ($self, $msg) = @_;
  $self->emit(message => $msg);
}

sub _legacy_dispatch_message {
  my ($self, $msg) = @_;
  my $event = $msg->{event};

  $event = "irc_$event" unless $event =~ /^(ctcp(reply)?|err)_/;
  warn "[$self->{debug_key}] === $event\n" if DEBUG == 2;
  $self->emit($event => $msg);
}

# Can be used in unittest to mock input data:
# $irc->_read($bytes);
sub _read {
  my $self = shift;

  no warnings 'utf8';
  $self->{buffer} .= Unicode::UTF8::decode_utf8($_[0], sub { $_[0] });

CHUNK:
  while ($self->{buffer} =~ s/^([^\015\012]+)[\015\012]//m) {
    warn "[$self->{debug_key}] >>> $1\n" if DEBUG;
    my $msg = $self->parser->parse($1);
    my $cmd = $msg->{command} or next CHUNK;
    $msg->{command} = $NUMERIC2NAME{$cmd} || IRC::Utils::numeric_to_name($cmd) || $cmd if $cmd =~ /^\d+$/;
    $msg->{event}   = lc $msg->{command};
    $self->_dispatch_message($msg);
  }
}

1;

=encoding utf8

=head1 NAME

Mojo::IRC - IRC Client for the Mojo IOLoop

=head1 SYNOPSIS

  my $irc = Mojo::IRC->new(
              nick => 'test123',
              user => 'my name',
              server => 'irc.perl.org:6667',
            );

  $irc->on(irc_join => sub {
    my($self, $message) = @_;
    warn "yay! i joined $message->{params}[0]";
  });

  $irc->on(irc_privmsg => sub {
    my($self, $message) = @_;
    say $message->{prefix}, " said: ", $message->{params}[1];
  });

  $irc->connect(sub {
    my($irc, $err) = @_;
    return warn $err if $err;
    $irc->write(join => '#mojo');
  });

  Mojo::IOLoop->start;

=head1 DESCRIPTION

L<Mojo::IRC> is a non-blocking IRC client using L<Mojo::IOLoop> from the
wonderful L<Mojolicious> framework.

It features IPv6 and TLS, with additional optional modules:
L<IO::Socket::IP> and L<IO::Socket::SSL>.

By default this module will only emit standard IRC events, but by
settings L</parser> to a custom object it will also emit CTCP events.
Example:

  my $irc = Mojo::IRC->new;
  $irc->parser(Parse::IRC->new(ctcp => 1);
  $irc->on(ctcp_action => sub {
    # ...
  });

It will also set up some default events: L</ctcp_ping>, L</ctcp_time>,
and L</ctcp_version>.

This class inherits from L<Mojo::EventEmitter>.

=head1 TESTING

The module L<Test::Mojo::IRC> is useful if you want to write tests without
having a running IRC server.

L<MOJO_IRC_OFFLINE> (from v0.20) is now DEPRECATED in favor of
L<Test::Mojo::IRC>.

=head1 EVENTS

=head2 close

  $self->on(close => sub { my ($self) = @_; });

Emitted once the connection to the server closes.

=head2 error

  $self->on(error => sub { my ($self, $err) = @_; });

Emitted once the stream emits an error.

=head2 message

  $self->on(message => sub { my ($self, $msg) = @_; });

Emitted when a new IRC message arrives. Will dispatch to a default handler,
which will again emit L</err_event_name> L</ctcp_event_name> and
L</irc_event_name> below.

Here is an example C<$msg>:

  {
    command  => "PRIVMSG",
    event    => "privmsg",
    params   => ["#convos", "hey!"],
    prefix   => "jan_henning",
    raw_line => ":jan_henning PRIVMSG #convos :hey",
  }

=head2 err_event_name

Events that start with "err_" are emitted when there is an IRC response that
indicates an error. See L<Mojo::IRC::Events> for sample events.

=head2 ctcp_event_name

Events that start with "ctcp_" are emitted if the L</parser> can understand
CTCP messages, and there is a CTCP response.

  $self->parser(Parse::IRC->new(ctcp => 1);

See L<Mojo::IRC::Events> for sample events.

=head2 irc_event_name

Events that start with "irc_" are emitted when there is a normal IRC response.
See L<Mojo::IRC::Events> for sample events.

=head1 ATTRIBUTES

=head2 connect_timeout

  $int = $self->connect_timeout;
  $self = $self->connect_timeout(60);

Maximum amount of time in seconds establishing a connection may take before
getting canceled, defaults to the value of the C<MOJO_IRC_CONNECT_TIMEOUT>
environment variable or 30.

=head2 ioloop

Holds an instance of L<Mojo::IOLoop>.

=head2 local_address

  $str = $self->local_address;
  $self = $self->local_address("10.20.30.40");

Local address to bind to. See L<Mojo::IOLoop::Client/local_address>.

=head2 name

The name of this IRC client. Defaults to "Mojo IRC".

=head2 nick

IRC nick name accessor. Default to L</user>.

=head2 parser

  $self = $self->parser($obj);
  $self = $self->parser(Parse::IRC->new(ctcp => 1));
  $obj = $self->parser;

Holds a L<Parse::IRC> object by default.

=head2 pass

Password for authentication

=head2 real_host

Will be set by L</irc_rpl_welcome>. Holds the actual hostname of the IRC
server that we are connected to.

=head2 server

Server name and, optionally, a port to connect to. Changing this while
connected to the IRC server will issue a reconnect.

=head2 server_settings

  $hash = $self->server_settings;

Holds information about the server. See
L<https://github.com/jhthorsen/mojo-irc/blob/master/t/ua-channel-users.t> for
example data structure.

Note that this attribute is EXPERIMENTAL and the structure of the values it
holds.

=head2 user

IRC username. Defaults to current logged in user or falls back to "anonymous".

=head2 tls

  $self->tls(undef) # disable (default)
  $self->tls({}) # enable

Default is "undef" which disables TLS. Setting this to an empty hash will
enable TLS and this module will load in default certs. It is also possible
to set custom cert/key:

  $self->tls({ cert => "/path/to/client.crt", key => ... })

This can be generated using

  # certtool --generate-privkey --outfile client.key
  # certtool --generate-self-signed --load-privkey client.key --outfile client.crt

To disable the verification of server certificates, the "insecure" option
can be set:

  $self->tls({insecure => 1});

=head1 METHODS

=head2 connect

  $self = $self->connect(\&callback);

Will log in to the IRC L</server> and call C<&callback>. The
C<&callback> will be called once connected or if connect fails. The second
argument will be an error message or empty string on success.

=head2 ctcp

  $str = $self->ctcp(@str);

This message will quote CTCP messages. Example:

  $self->write(PRIVMSG => nickname => $self->ctcp(TIME => time));

The code above will write this message to IRC server:

  PRIVMSG nickname :\001TIME 1393006707\001

=head2 disconnect

  $self->disconnect(\&callback);

Will disconnect form the server and run the callback once it is done.

=head2 new

  $self = Mojo::IRC->new(%attrs);

Object constructor.

=head2 register_default_event_handlers

  $self->register_default_event_handlers;

This method sets up the default L</DEFAULT EVENT HANDLERS> unless someone has
already subscribed to the event.

=head2 write

  $self->write(@str, \&callback);

This method writes a message to the IRC server. C<@str> will be concatenated
with " " and "\r\n" will be appended. C<&callback> is called once the message is
delivered over the stream. The second argument to the callback will be
an error message: Empty string on success and a description on error.

=head2 write_p

  $promise = $self->write_p(@str);

Like L</"write">, but returns a L<Mojo::Promise> instead of taking a callback.
The promise will be resolved on success, or rejected with the error message on
error.

=head1 DEFAULT EVENT HANDLERS

=head2 ctcp_ping

Will respond to the sender with the difference in time.

  Ping reply from $sender: 0.53 second(s)

=head2 ctcp_time

Will respond to the sender with the current localtime. Example:

  TIME Fri Feb 21 18:56:50 2014

NOTE! The localtime format may change.

=head2 ctcp_version

Will respond to the sender with:

  VERSION Mojo-IRC $VERSION

NOTE! Additional information may be added later on.

=head2 irc_nick

Used to update the L</nick> attribute when the nick has changed.

=head2 irc_notice

Responds to the server with "QUOTE PASS ..." if the notice contains "Ident
broken...QUOTE PASS...".

=head2 irc_ping

Responds to the server with "PONG ...".

=head2 irc_rpl_isupport

Used to populate L</server_settings> with information about the server.

=head2 irc_rpl_welcome

Used to get the hostname of the server. Will also set up automatic PING
requests to prevent timeout and update the L</nick> attribute.

=head2 err_nicknameinuse

This handler will add "_" to the failed nick before trying to register again.

=head1 COPYRIGHT

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 AUTHOR

Marcus Ramberg - C<mramberg@cpan.org>

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut
